src/libraries/System.Linq/src/System/Linq/Sum.cs

```diff
@@ -18,9 +18,19 @@ public static int Sum(this IEnumerable<int> source)
             int sum = 0;
             checked
             {
-                foreach (int v in source)
+                if (source is int[] array)
```

Comment:
I guess it should be `is IList<int>` instead to handle more cases.
Such optimizations used to be rejected due to performance penalties because of additional casts (when you improve your case and regress others). But maybe it's not a case anymore thanks to cast caching feature?
Perhaps, makes sense to include other types of input into your benchmark.

Reply:
I don't think that the performance penalty for the single extra cast is high. The benchmark shows a significant performance improvement, also for short arrays.

I don't think it should only test for `IList<int>`. The indexer of the array returns the element by reference, not by value. This improves the performance. I suggest to implement it **also** on 'IList<T>' (where T is an int, int?, double, double?, long, long?, decimal, decimal?).  This would improve the performance with about a factor 3. Less than on an array but that is because of the slower indexer.

The code for the Sum() on int would look like:
```
public static int Sum(this IEnumerable<int> source) {
    if (source == null) {
        ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
    }

    int sum = 0;
    checked {
        if (source is int[] array) {
            for (int i = 0; i < array.Length; i++) {
                sum += array[i];
            }

        } else if (source is IList<int> list) {
            for (int i = 0; i < list.Count; i++) {
                sum += list[i];
            }
        } 
        else 
        {
            foreach (int v in source) {
                sum += v;
            }
        }
    }

    return sum;
}
```

The benchmark which I used to validate this:
```
[MemoryDiagnoser]
public class ListOfIntBM {

    private List<int> _list;

    [Params(10, 100, 1000, 10000)]
    public int Size { get; set; }

    [GlobalSetup]
    public void Setup() {
        _list = new List<int>(Size);
        for (int i = 0; i < _list.Count; i++) {
            _list[i] = i;
        }
    }

    [Benchmark]
    public int Old() => System.Linq.Enumerable.Sum(_list);

    [Benchmark(Baseline = true)]
    public int New() => System.LinqNew.Enumerable.Sum(_list);
}
```

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18362
AMD Ryzen 9 3900X, 1 CPU, 24 logical and 12 physical cores
.NET Core SDK=3.1.200-preview-015002
  [Host]     : .NET Core 3.1.2 (CoreCLR 4.700.20.6602, CoreFX 4.700.20.6702), X64 RyuJIT
  DefaultJob : .NET Core 3.1.2 (CoreCLR 4.700.20.6602, CoreFX 4.700.20.6702), X64 RyuJIT


```
| Method |  Size |      Mean |     Error |    StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |------ |----------:|----------:|----------:|------:|--------:|-------:|------:|------:|----------:|
|    **Old** |    **10** | **16.066 ns** | **0.3499 ns** | **0.3889 ns** |  **2.62** |    **0.09** | **0.0048** |     **-** |     **-** |      **40 B** |
|    New |    10 |  6.126 ns | 0.1388 ns | 0.1299 ns |  1.00 |    0.00 |      - |     - |     - |         - |
|        |       |           |           |           |       |         |        |       |       |           |
|    **Old** |   **100** | **16.033 ns** | **0.3376 ns** | **0.3467 ns** |  **2.73** |    **0.06** | **0.0048** |     **-** |     **-** |      **40 B** |
|    New |   100 |  5.891 ns | 0.0650 ns | 0.0608 ns |  1.00 |    0.00 |      - |     - |     - |         - |
|        |       |           |           |           |       |         |        |       |       |           |
|    **Old** |  **1000** | **16.399 ns** | **0.3449 ns** | **0.3691 ns** |  **3.06** |    **0.12** | **0.0048** |     **-** |     **-** |      **40 B** |
|    New |  1000 |  5.358 ns | 0.1334 ns | 0.1370 ns |  1.00 |    0.00 |      - |     - |     - |         - |
|        |       |           |           |           |       |         |        |       |       |           |
|    **Old** | **10000** | **16.729 ns** | **0.1763 ns** | **0.1472 ns** |  **3.15** |    **0.11** | **0.0048** |     **-** |     **-** |      **40 B** |
|    New | 10000 |  5.316 ns | 0.1323 ns | 0.1767 ns |  1.00 |    0.00 |      - |     - |     - |         - |


I will implement the `IList<T>` solution in the code if we can agree on this.